contracts/minter-suite/Minters/MinterDAExpRefund/MinterDAExpRefundV0.sol

```diff
@@ -0,0 +1,997 @@
+        // if the price is base price, the auction is valid and may be claimed
+        // calculate the artist and admin revenues (no check requuired)
+        uint256 netRevenues = _projectConfig.numRefundableInvocations * _price;
+        // INTERACTIONS
+        _splitETHRevenues(_projectId, netRevenues);
+        emit ArtistAndAdminRevenuesWithdrawn(_projectId);
+    }
+
+    /**
+     * @notice Purchases a token from project `_projectId`.
+     * @param _projectId Project ID to mint a token on.
+     * @return tokenId Token ID of minted token
+     */
+    function purchase(uint256 _projectId)
+        external
+        payable
+        returns (uint256 tokenId)
+    {
+        tokenId = purchaseTo_do6(msg.sender, _projectId);
+        return tokenId;
+    }
+
+    /**
+     * @notice gas-optimized version of purchase(uint256).
+     */
+    function purchase_H4M(uint256 _projectId)
+        external
+        payable
+        returns (uint256 tokenId)
+    {
+        tokenId = purchaseTo_do6(msg.sender, _projectId);
+        return tokenId;
+    }
+
+    /**
+     * @notice Purchases a token from project `_projectId` and sets
+     * the token's owner to `_to`.
+     * @param _to Address to be the new token's owner.
+     * @param _projectId Project ID to mint a token on.
+     * @return tokenId Token ID of minted token
+     */
+    function purchaseTo(address _to, uint256 _projectId)
+        external
+        payable
+        returns (uint256 tokenId)
+    {
+        return purchaseTo_do6(_to, _projectId);
+    }
+
+    /**
+     * @notice gas-optimized version of purchaseTo(address, uint256).
+     */
+    function purchaseTo_do6(address _to, uint256 _projectId)
+        public
+        payable
+        nonReentrant
+        returns (uint256 tokenId)
+    {
+        // CHECKS
+        ProjectConfig storage _projectConfig = projectConfig[_projectId];
+
+        // Note that `maxHasBeenInvoked` is only checked here to reduce gas
+        // consumption after a project has been fully minted.
+        // `_projectConfig.maxHasBeenInvoked` is locally cached during every
+        // purchase to reduce gas consumption and enable recording of sellout
+        // price, but if not in sync with the core contract's value,
+        // the core contract also enforces its own max invocation check during
+        // minting.
+        require(
+            !_projectConfig.maxHasBeenInvoked,
+            "Maximum number of invocations reached"
+        );
+
+        // _getPrice reverts if auction is unconfigured or has not started
+        uint256 currentPriceInWei = _getPrice(_projectId);
+
+        // EFFECTS
+        // update the purchaser's receipt and require sufficient net payment
+        Receipt storage _receipt = receipts[msg.sender][_projectId];
+        _receipt.netPaid += msg.value;
+        _receipt.numPurchased++;
+        require(
+            _receipt.netPaid >= _receipt.numPurchased * currentPriceInWei,
```

Comment:
Revert invalidates all state changes in a given call, so it would not lead to an inaccurate receipt here.
I like the readability of the current code just because it is very easy to rationalize "require that the user paid >= price*quantity".

That being said, we are mixing a check after some effects. However, because the effects are all very local and easy to see that no reentrancy can occur, there is no cause for concern here.

There would be a way to improve gas efficiency here, but it would lead to less readable code, so 🤷 seems not worth it.

I think this is okay as-is, but I'll let you resolve or continue the discussion @mchrupcala

Reply:
Nope I 100% agree with you re: readability -- especially for the sensitive nature of the refund's math, I agree that it's great to prioritize readability over gas-savings here. Thanks for confirming that a revert would invalidate any state changes. Looks great to me!